### PR TITLE
MRG: decrease memory usage of rERPs

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
as discussed w/, and mostly written by, @dengemann

Cuts memory usage by approximately a factor of 2 in typical usage cases on our machine. Maybe now the example will run on @mainakjas' laptop!

Partially adresses #2331

### CODE REVIEW

File: mne/stats/regression.py [hunk 1]
```diff
@@ -279,21 +279,24 @@ def solver(X, Y):
...
         lags = np.zeros(n_lags, dtype=float)
-
         if cond in event_id:  # for binary predictors
-            ids = ([event_id[cond]] if isinstance(event_id[cond], int)
+            ids = ([event_id[cond]]
+                   if isinstance(event_id[cond], int)
                    else event_id[cond])
             samples[events[np.in1d(events[:, 2], ids), 0] + int(tmin_)] = 1
-            cond_length[cond] = sum(np.in1d(events[:, 2], ids))
+            cond_length[cond] = np.sum(np.in1d(events[:, 2], ids))
```
Comment (agramfort) [inline mne/stats/regression.py]: I missed this one. Use in1d from fixes for old numpy.

File: mne/stats/regression.py [hunk 2]
```diff
@@ -306,24 +309,21 @@ def solver(X, Y):
...
                                         info, tstep)
         for t0, t1 in inds:
             has_val[t0:t1] = False
 
-    X = big_arr[:, has_val].T
-    Y = data[:, has_val]
-
     # solve linear system
-    coefs = solver(X, Y)
+    coefs = solver(X[has_val], data[:, has_val])
```
Comment (agramfort) [inline mne/stats/regression.py]: indexing with a boolean will make a copy FYI but it's more a remark

**Comment by agramfort** [comment]
besides LGTM

File: mne/stats/regression.py (hunk 2 above)
Comment (jona-sassenhagen) [inline mne/stats/regression.py]: No that's actually a major problem this PR should have caught .. something I missed when trying to translate Denis' script. So is it better if I overwrite it with itself?

```
X, data = X[has_val], data[:, has_val]
coefs = solver(X, data)
```

Comment (agramfort) [inline mne/stats/regression.py]: that's doing the same. When you do

X = X[has_val]

a new array is allocated for X[has_val] and the output is stored in X.
So at 1 point you have the 2 copies in memory before X is garbage
collected.

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: IIRC Denis' profiling showed copying data was an important contributor to memory usage. Should I try np.take or np.compress?

Comment (agramfort) [inline mne/stats/regression.py]: unless you use slices you'll make copies. np.take is faster for access on
1st axis but does not limit memory.

The big plus in memory will be if you compute online X.T * X and X.T y as
suggested by @nsmith

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: On my system, boolean indexing is actually faster by about 10%?

```
%timeit for x in range(10): X.take(has_val, axis=0), data.take(has_val, axis=1)
1 loops, best of 3: 59.5 s per loop
%timeit for x in range(10): X[has_val, :], data[:, has_val]
1 loops, best of 3: 53.2 s per loop
```

**Comment by jona-sassenhagen** [comment]
Either way, I think this is done for now. @dengemann can you take a look, and merge if it's okay?

**Comment by agramfort** [comment]
merged by rebase after a tiny cosmit.

**Pull request merged**